api/src/Entity/ContentNode/ChecklistNode.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace App\Entity\ContentNode;
+
+use ApiPlatform\Metadata\ApiProperty;
+use ApiPlatform\Metadata\ApiResource;
+use ApiPlatform\Metadata\Delete;
+use ApiPlatform\Metadata\Get;
+use ApiPlatform\Metadata\GetCollection;
+use ApiPlatform\Metadata\Patch;
+use ApiPlatform\Metadata\Post;
+use App\Entity\Checklist;
+use App\Entity\ChecklistItem;
+use App\Entity\ContentNode;
+use App\Repository\ChecklistNodeRepository;
+use App\State\ContentNode\ChecklistNodePersistProcessor;
+use App\Util\EntityMap;
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\Serializer\Annotation\Groups;
+
+#[ApiResource(
+    operations: [
+        new Get(
+            security: 'is_granted("CAMP_COLLABORATOR", object) or is_granted("CAMP_IS_PROTOTYPE", object)'
+        ),
+        new Patch(
+            processor: ChecklistNodePersistProcessor::class,
+            denormalizationContext: ['groups' => ['write', 'update']],
+            security: 'is_granted("CAMP_MEMBER", object) or is_granted("CAMP_MANAGER", object)',
+            validationContext: ['groups' => ['Default', 'update']]
+        ),
+        new Delete(
+            security: '(is_granted("CAMP_MEMBER", object) or is_granted("CAMP_MANAGER", object)) and object.parent !== null'
+        ),
+        new GetCollection(
+            security: 'is_authenticated()'
+        ),
+        new Post(
+            processor: ChecklistNodePersistProcessor::class,
+            denormalizationContext: ['groups' => ['write', 'create']],
+            securityPostDenormalize: 'is_granted("CAMP_MEMBER", object) or is_granted("CAMP_MANAGER", object)',
+            validationContext: ['groups' => ['Default', 'create']],
+        ),
+    ],
+    denormalizationContext: ['groups' => ['write']],
+    normalizationContext: ['groups' => ['read']],
+    routePrefix: '/content_node'
+)]
+#[ORM\Entity(repositoryClass: ChecklistNodeRepository::class)]
+class ChecklistNode extends ContentNode {
+    /**
+     * The content types that are most likely to be useful for planning programme of this category.
+     */
+    #[ApiProperty(example: '["/checklist_items/1a2b3c4d"]')]
+    #[Groups(['read'])]
+    #[ORM\ManyToMany(targetEntity: ChecklistItem::class, inversedBy: 'checklistNodes')]
+    #[ORM\JoinTable(name: 'checklistnode_checklistitem')]
+    #[ORM\JoinColumn(name: 'checklistnode_id', referencedColumnName: 'id', onDelete: 'CASCADE')]
+    #[ORM\InverseJoinColumn(name: 'checklistitem_id', referencedColumnName: 'id', onDelete: 'CASCADE')]
+    #[ORM\OrderBy(['position' => 'ASC'])]
+    public Collection $checklistItems;
+
+    #[ApiProperty(example: '["1a2b3c4d"]')]
+    #[Groups(['write'])]
+    public ?array $addChecklistItemIds = [];
+
+    #[ApiProperty(example: '["1a2b3c4d"]')]
+    #[Groups(['write'])]
+    public ?array $removeChecklistItemIds = [];
```

Comment:
What is this used for?

Reply:
`[ChecklistNode]<*-*>[ChecklistItem]` is an N to M relationship, which only exists in Doctrine. If you read a `[ChecklistNode]` from the API, you will also get all selected `[ChecklistItems]`.

On the GUI you will select and deselect individual ChecklistItems.

I have implemented exactly this functionality on the API. You do not have to send all ChecklistItems with a PATCH to a ChecklistNode. Instead, you can transmit which ChecklistItems should be added or removed.

**Add ChecklistItem:**
```
PATCH /content_node/checklist_node/{id} 
{
  addChecklistItemIds: [ 'a1b2c3d4' ]
}
```

**Remove ChecklistItem:**
```
PATCH /content_node/checklist_node/{id} 
{
  removeChecklistItemIds: [ 'a1b2c3d4' ]
}
```

Reply:
good catch
slowly on the way to merge patch json.

we normally would use IRI to identifiy the CheckListItems...but maybe this is more difficult to extract the ids later.
Maybe if you had a transient field, api platform woud fetch the Items directly for you?